x/bep3/types/msg.go

```diff
@@ -0,0 +1,305 @@
+
+// Type is the name of MsgCreateHTLT
+func (msg MsgCreateHTLT) Type() string { return Htlt }
+
+// String prints the MsgCreateHTLT
+func (msg MsgCreateHTLT) String() string {
+	return fmt.Sprintf("HTLT{%v#%v#%v#%v#%v#%v#%v#%v#%v#%v}",
+		msg.From, msg.To, msg.RecipientOtherChain, msg.SenderOtherChain, msg.RandomNumberHash,
+		msg.Timestamp, msg.Amount, msg.ExpectedIncome, msg.HeightSpan, msg.CrossChain)
+}
+
+// GetInvolvedAddresses gets the addresses involved in a MsgCreateHTLT
+func (msg MsgCreateHTLT) GetInvolvedAddresses() []sdk.AccAddress {
+	return append(msg.GetSigners(), AtomicSwapCoinsAccAddr)
+}
+
+// GetSigners gets the signers of a MsgCreateHTLT
+func (msg MsgCreateHTLT) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{msg.From}
+}
+
+// ValidateBasic validates the MsgCreateHTLT
+func (msg MsgCreateHTLT) ValidateBasic() sdk.Error {
+	if len(msg.From) != AddrByteCount {
+		return sdk.ErrInternal(fmt.Sprintf("the expected address length is %d, actual length is %d", AddrByteCount, len(msg.From)))
+	}
+	if len(msg.To) != AddrByteCount {
+		return sdk.ErrInternal(fmt.Sprintf("the expected address length is %d, actual length is %d", AddrByteCount, len(msg.To)))
+	}
+	if !msg.CrossChain && len(msg.RecipientOtherChain) != 0 {
+		return sdk.ErrInternal(fmt.Sprintf("must leave recipient address on other chain to empty for single chain swap"))
+	}
+	if !msg.CrossChain && len(msg.SenderOtherChain) != 0 {
+		return sdk.ErrInternal(fmt.Sprintf("must leave sender address on other chain to empty for single chain swap"))
+	}
+	if msg.CrossChain && len(msg.RecipientOtherChain) == 0 {
+		return sdk.ErrInternal(fmt.Sprintf("missing recipient address on other chain for cross chain swap"))
+	}
+	if len(msg.RecipientOtherChain) > MaxOtherChainAddrLength {
+		return sdk.ErrInternal(fmt.Sprintf("the length of recipient address on other chain should be less than %d", MaxOtherChainAddrLength))
+	}
+	if len(msg.SenderOtherChain) > MaxOtherChainAddrLength {
+		return sdk.ErrInternal(fmt.Sprintf("the length of sender address on other chain should be less than %d", MaxOtherChainAddrLength))
+	}
+	if len(msg.ExpectedIncome) > MaxExpectedIncomeLength {
+		return sdk.ErrInternal(fmt.Sprintf("the length of expected income should be less than %d", MaxExpectedIncomeLength))
+	}
+	if len(msg.RandomNumberHash) != RandomNumberHashLength {
+		return sdk.ErrInternal(fmt.Sprintf("the length of random number hash should be %d", RandomNumberHashLength))
+	}
+	if !msg.Amount.IsAllPositive() {
+		return sdk.ErrInternal(fmt.Sprintf("the swapped out coin must be positive"))
+	}
+	return nil
+}
+
+// GetSignBytes gets the sign bytes of a MsgCreateHTLT
+func (msg MsgCreateHTLT) GetSignBytes() []byte {
+	b, err := json.Marshal(msg)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+// MsgDepositHTLT defines an HTLT deposit
+type MsgDepositHTLT struct {
+	From   sdk.AccAddress `json:"from"`
+	SwapID string         `json:"swap_id"`
```

Comment:
Is there a reason `SwapID` is `string` compared to `[32]byte`? BEP3 lists it as 32 bytes.
The MarshalJSON method will take care of encoding it to hex automatically when a msg is encoded to json.

Reply:
Included in https://github.com/Kava-Labs/kava/pull/370/commits/d0514831d6734a722e3402739ed07fa44dfbbff3